sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeJsonExprs.scala

```diff
@@ -28,10 +28,36 @@ import org.apache.spark.sql.types.{ArrayType, StructType}
  * The optimization includes:
  * 1. JsonToStructs(StructsToJson(child)) => child.
  * 2. Prune unnecessary columns from GetStructField/GetArrayStructFields + JsonToStructs.
+ * 3  struct(from_json.col1, from_json.col2, from_json.col3...) => struct(from_json)
  */
 object OptimizeJsonExprs extends Rule[LogicalPlan] {
   override def apply(plan: LogicalPlan): LogicalPlan = plan transform {
     case p => p.transformExpressions {
+
+      case c: CreateNamedStruct
+        if c.valExprs.forall(v => v.isInstanceOf[GetStructField] &&
+          v.asInstanceOf[GetStructField].child.isInstanceOf[JsonToStructs]) =>
+        val jsonToStructs = c.valExprs.map(_.children(0))
+        val semanticEqual = jsonToStructs.tail.forall(jsonToStructs.head.semanticEquals(_))
+        val sameFieldName = c.names.zip(c.valExprs).forall {
+          case (name, valExpr: GetStructField) =>
+            name.toString == valExpr.childSchema(valExpr.ordinal).name
```

Comment:
Does this work correctly if multiple values refer to the same ordinal?
```
scala> sql("""SELECT from_json('{"a":1, "b":0.8}', 'a INT, b DOUBLE') v""").createOrReplaceTempView("t")
scala> sql("select named_struct('a', t.v.a, 'a', t.v.a) from t").explain(true)
== Parsed Logical Plan ==
'Project [unresolvedalias('named_struct(a, 't.v.a, a, 't.v.a), None)]
+- 'UnresolvedRelation [t], [], false

== Analyzed Logical Plan ==
named_struct(a, v.a AS `a`, a, v.a AS `a`): struct<a:int,a:int>
Project [named_struct(a, v#128.a, a, v#128.a) AS named_struct(a, v.a AS `a`, a, v.a AS `a`)#133]
+- SubqueryAlias t
   +- Project [from_json(StructField(a,IntegerType,true), StructField(b,DoubleType,true), {"a":1, "b":0.8}, Some(Asia/Tokyo)) AS v#128]
      +- OneRowRelation

== Optimized Logical Plan ==
Project [[1,1] AS named_struct(a, v.a AS `a`, a, v.a AS `a`)#133]
+- OneRowRelation

== Physical Plan ==
*(1) Project [[1,1] AS named_struct(a, v.a AS `a`, a, v.a AS `a`)#133]
+- *(1) Scan OneRowRelation[]
```

Reply:
Oh, this is kind of corner case. We cannot simplify this case. Added test case.